<a href="https://colab.research.google.com/github/bcury/Machine-Learning-Projects/blob/main/ML_Regress%C3%A3o_Previs%C3%A3o_DecisionTree_RegressaoLinear_RandomForest.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

- Base plano de saúde

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express  as px

In [3]:
base_plano_saude = pd.read_csv('/content/plano_saude.csv')

In [4]:
base_plano_saude

idade  custo
0     18    871
1     23   1132
2     28   1242
3     33   1356
4     38   1488
5     43   1638
6     48   2130
7     53   2454
8     58   3066
9     63   4090

In [5]:
x_plano_saude = base_plano_saude['idade'].values

In [6]:
y_plano_saude = base_plano_saude['custo'].values

In [7]:
x_plano_saude, y_plano_saude

(array([18, 23, 28, 33, 38, 43, 48, 53, 58, 63]),
 array([ 871, 1132, 1242, 1356, 1488, 1638, 2130, 2454, 3066, 4090]))

In [8]:
np.corrcoef(x_plano_saude, y_plano_saude)

array([[1.        , 0.93091958],
       [0.93091958, 1.        ]])

In [9]:
x_plano_saude.shape

(10,)

In [10]:
x_plano_saude = x_plano_saude.reshape(-1,1)

In [11]:
x_plano_saude.shape

(10, 1)

In [12]:
x_plano_saude

array([[18],
       [23],
       [28],
       [33],
       [38],
       [43],
       [48],
       [53],
       [58],
       [63]])

In [13]:
y_plano_saude

array([ 871, 1132, 1242, 1356, 1488, 1638, 2130, 2454, 3066, 4090])

In [14]:
y_plano_saude.shape

(10,)

In [15]:
from sklearn.model_selection import train_test_split

In [16]:
x_plano_saude_treino, x_plano_saude_teste, y_plano_saude_treino, y_plano_saude_teste = train_test_split(x_plano_saude, y_plano_saude, test_size=0.25, random_state=1)

**Regressao Linear**

In [17]:
from sklearn.linear_model import LinearRegression

In [18]:
regressao_plano_saude = LinearRegression()

In [19]:
regressao_plano_saude.fit(x_plano_saude_treino, y_plano_saude_treino)

LinearRegression()

In [20]:
previsao = regressao_plano_saude.predict(x_plano_saude_teste)

In [21]:
previsao

array([1213.92307692, 2967.69230769, 2216.07692308])

In [22]:
y_plano_saude_teste

array([1242, 4090, 2130])

In [23]:
regressao_plano_saude.score(x_plano_saude_treino, y_plano_saude_treino)

0.9041878543873897

In [24]:
regressao_plano_saude.score(x_plano_saude_teste, y_plano_saude_teste)

0.7014957781587308

modelo sofreu de overfitting, excelnte acuracia de treino e razoavel acuracia de teste

**Decision Tree**

In [25]:
x_plano_saude, y_plano_saude

(array([[18],
        [23],
        [28],
        [33],
        [38],
        [43],
        [48],
        [53],
        [58],
        [63]]),
 array([ 871, 1132, 1242, 1356, 1488, 1638, 2130, 2454, 3066, 4090]))

In [26]:
from sklearn.tree import DecisionTreeRegressor

In [45]:
regressor_arvore = DecisionTreeRegressor(criterion='squared_error')

In [46]:
regressor_arvore.fit(x_plano_saude_treino, y_plano_saude_treino)

DecisionTreeRegressor()

In [47]:
previsoes = regressor_arvore.predict(x_plano_saude_teste)

In [48]:
previsoes

array([1132., 3066., 1638.])

In [49]:
y_plano_saude_teste

array([1242, 4090, 2130])

In [50]:
regressor_arvore.score(x_plano_saude_treino,y_plano_saude_treino)

1.0

In [51]:
regressor_arvore.score(x_plano_saude_teste, y_plano_saude_teste)

0.6932623868556675

overfitting também aqui, "apredeu muito o treino"

o que podemos fazer é colher mais dadospara o dataset, assim o modelo nao deve aprender tanto no treino e vamos balencear bem com os testes

mas podemos fazer previsoes:

In [52]:
regressor_arvore.predict(np.array([[22]])) #idade 22, o plano vai custar:

array([1132.])

o que bate com a base real, na base real uma pessoa de 23 anos paga o plano com o valor de 1132 também,ou seja, faz sentido a previsão para a idade de 22 anos

In [53]:
regressor_arvore.predict(np.array([[55]])) #idade 55, o plano vai custar:

array([2454.])

aqui também!!! corresponde ao mesmo valor pago por uma pessoa de 53 anos, ou seja, uma bela previsão

**Random Forest**

In [55]:
base_nova = pd.read_csv('/content/plano_saude2.csv')

In [56]:
base_nova

idade  custo
0     18    470
1     23    520
2     28    630
3     33    830
4     38   1150
5     43   1530
6     48   2040
7     53   3080
8     58   5100
9     63  10100

In [57]:
x_plano_saude_novo = base_nova.iloc[:,0].values
y_plano_saude_novo = base_nova.iloc[:,1].values

In [58]:
x_plano_saude_novo

array([18, 23, 28, 33, 38, 43, 48, 53, 58, 63])

In [66]:
x_plano_saude_novo = x_plano_saude_novo.reshape(-1,1)

In [59]:
y_plano_saude_novo

array([  470,   520,   630,   830,  1150,  1530,  2040,  3080,  5100,
       10100])

In [67]:
from sklearn.model_selection import train_test_split

In [68]:
x_plano_saude_novo_treino,x_Plano_saude_novo_teste, y_plano_saude_novo_treino, y_plano_saude_novo_teste = train_test_split( x_plano_saude_novo, y_plano_saude_novo, test_size=0.25, random_state=1)

In [69]:
from sklearn.ensemble import RandomForestRegressor

In [70]:
regressao_rf =  RandomForestRegressor(n_estimators=10,criterion='mse')

In [71]:
regressao_rf.fit(x_plano_saude_novo_treino, y_plano_saude_novo_treino)

/usr/local/lib/python3.7/dist-packages/sklearn/ensemble/_forest.py:392: FutureWarning:

Criterion 'mse' was deprecated in v1.0 and will be removed in version 1.2. Use `criterion='squared_error'` which is equivalent.



RandomForestRegressor(criterion='mse', n_estimators=10)

In [72]:
predict = regressao_rf.predict(x_Plano_saude_novo_teste)

In [73]:
predict

array([ 500., 5100., 2150.])

In [74]:
y_plano_saude_novo_teste

array([  630, 10100,  2040])

In [75]:
regressao_rf.score(x_plano_saude_novo_treino, y_plano_saude_novo_treino)

0.9578353355658336

96% de acurácia!!

In [78]:
regressao_rf.predict(np.array([[55]]))# idade 55, o plano vai custar:

array([3888.])

faz muito sentido essa previsão comparando com a base de dados